app/src/org/commcare/dalvik/activities/CommCareHomeActivity.java

```diff
@@ -786,6 +627,140 @@ else if(resultCode == RESULT_OK) {
 
     }
 
+    /**
+     * Process user returning home from the form entry activity.
+     * Triggers form submission cycle, cleans up some session state.
+     *
+     * @param resultCode exit code of form entry activity
+     * @param intent points to form instance
+     */
+    private void processReturnFromFormEntry(int resultCode, Intent intent) {
+        // TODO: We might need to load this from serialized state?
+        AndroidSessionWrapper currentState = CommCareApplication._().getCurrentSessionWrapper();
+
+        // This is the state we were in when we _Started_ form entry
+        FormRecord current = currentState.getFormRecord();
+
+        if (current == null) {
+            // somehow we lost the form record for the current session
+            // TODO: how should this be handled? -- PLM
+            Toast.makeText(this,
+                    "Error while trying to save the form!",
+                    Toast.LENGTH_LONG);
+            Logger.log(AndroidLogger.TYPE_ERROR_WORKFLOW,
+                    "Form Entry couldn't save because of corrupt state.");
+            clearSessionAndExit(currentState);
+        }
+
+        // TODO: This should be the default unless we're in some "Uninit" or "incomplete" state
+        if (FormRecord.STATUS_COMPLETE.equals(current.getStatus()) ||
+                FormRecord.STATUS_SAVED.equals(current.getStatus()) ||
+                FormRecord.STATUS_UNSENT.equals(current.getStatus())) {
+            // Viewing an old form, so don't change the historical record
+            // regardless of the exit code
+            currentState.reset();
+            if (wasExternal) {
+                this.finish();
```

Comment:
Essentially there were two code paths out of the MODEL_RESULT case before (Listed below). easiest thing to do would be to have the new method have a return code that either returned out of the method, or fell through to the next fetch. Alternative would be to more deliberately encode the potential outcomes for all of these activity returns and follow that path

1) fallthrough (break, or simply continue executing) - These exits would jump into a session. This was the return path if, say, the user canceled form entry, or a new session frame was placed on the stack and the user should be directed to it. These paths would result in the "startNextFetch" method being called at the bottom of the activity result
2) exit (paths previously ending in 'return') - these exits would not launch a new activity. They would either launch the Home Activity, or (if "finish" was called) they would leave commcare entirely (if CommCare was launched in response to an external app)